core/arch/opnd_shared.c

```diff
@@ -1806,6 +1806,41 @@ reg_set_value_priv(reg_id_t reg, priv_mcontext_t *mc, reg_t value)
     *(reg_t *)((byte *)mc + opnd_get_reg_mcontext_offs(reg)) = value;
 }
 
+bool
+reg_set_value_ex_priv(reg_id_t reg, priv_mcontext_t *mc, byte *val_buf, size_t size)
+{
+#ifdef X86
+    if (reg == REG_NULL)
+        return true;
+
+    opnd_size_t reg_size = reg_get_size(reg);
+
+    if (reg_size == OPSZ_NA)
+        return false;
+
+    if (opnd_size_in_bytes(reg_size) < size)
+        return false;
+
+    dr_zmm_t *simd = (dr_zmm_t *)((byte *)mc + SIMD_OFFSET);
+
+    if (reg_is_gpr(reg)) {
+        byte *reg_val_addr = ((byte *)mc + opnd_get_reg_mcontext_offs(reg));
+        memcpy(reg_val_addr, val_buf, size);
+    } else if (reg >= DR_REG_START_XMM && reg <= DR_REG_STOP_XMM) {
+        memcpy(&(simd[reg - DR_REG_START_XMM]), val_buf, size);
```

Comment:
I feel like for these fixed sizes there can be a CLIENT_ASSERT(size == XMM_REG_SIZE, "size does not match simd register."). Ditto below.

Reply:
Thanks for your comment. What if you want to set only the first 8 bytes of a register ?

Reply:
Not sure if we should support partial registers here. I feel like this would rather lead to bugs than be helpful. Why would we need this? If we were to support this here, we'd have to start supporting "insert" here as well, i.e. why only supporting the "lower 8 bytes". I think the function is less fragile if only full registers are supported. But it should get pointed out in the API above, similar to "no-sub-registers" in reg_set_value. WDYT

Reply:
Yep, I see what you mean - I will add the asserts.

Reply:
Then on the other hand, the size is not needed at all, just a comment that the caller has to make sure that val_buf has 'size' readable bytes?

Reply:
size or not up to your judgment

Reply:
Done. I decided to leave the inclusion of the size to better ensure that the user has set up the correct buffer size ( and perhaps aid in preventing future overflows). Let me know if you want it to be removed.